### DISCUSSION TITLE
What is nushell way for bash's || between 2 commands?

### DISCUSSION THREAD

**Discussion by avkonst** [nushell/nushell#6667] [Q&A]
I am rewriting some of the bash scripts in nushell to see how far I can get with it, and I noticed I frequently need to do the thing which I did in bash:
```
external_command1 || external_command2
```
which is do the first, and if it exits negatively, do the second.

What is the nu way to achieve the same?

**Comment by fdncred** [comment]
I'd try semicolon to see if it works, like: `external_command1 ; external_command2`

**Comment by avkonst** [comment]
It does not. Semicolon seems has no meaning and an effect.

**Comment by fdncred** [comment]
Maybe it's only for internal commands. I believe the semicolon is supposed to execute the first command and if it has a 0 exit code, then execute the second command. I'm not familiar with || on bash so I'm not sure what it's supposed to do.

**Comment by avkonst** [comment]
`||` executes the first command and if it exits with non-zero, executes the second. If the first exits with zero, the second is not executed

**Comment by fdncred** [comment]
oh, thanks for explaining. I don't think nushell has anything like that yet. You may be able to hack together a solution by looking at `$env.LAST_EXIT_CODE` though.

**Comment by theAkito** [comment]
According to some official Nu blog post, Nu has short-circuit `||`. So, using `||` should be quite the same, but the command on the left has to return a clean `false`, I guess.

**Comment by cmpute** [comment]
To test against exit codes, this is what I got to work:
```
if (python -c "print('cmd1'); exit(-1)" | complete).exit_code != 0 { python -c "print('command2')" } 
```

Unfortunately the stdout of the first command will be captured by `complete`, which is less ideal.

I think it will be great if there is a builtin command to test the exit code. For example support something like `python -c "exit(-1)" | test`  which returns false, then you can use a syntax like `(cmd1 | test) || (cmd2 | test)` and even `(cmd1 | test) && (cmd2 | test)`

**Comment by jasha-hrp** [comment]
I'll add that bash's `&&` operator is also very convenient. In bash:
- `foo || bar`: Run `foo`, then run `bar` if `foo` failed.
- `foo && bar`: Run `foo`, then run `bar` if `foo` succeeded.
These can be chained, e.g. `foo && bar || baz` would run `foo`, then run `bar` if `foo` failed. `baz` gets run if either of `foo` or `bar` fail.

Nushell lacks concise "short-circuiting" command chaining capabilities like this as far as I can tell. You have to do something very verbose using `if`-statements and use `$env.LAST_EXIT_CODE` or `complete(...).exit_code == 0`

**Reply by jasha-hrp** [reply]
Though it does seem that nushell's `and` and `or` cover similar bases for commands that return `bool`. If `foo` and `bar` return bool then you can do:
- `(foo) or (bar)` (equivalent to bash's `foo || bar`)
- `(foo) and (bar)` (equivalent to bash's `foo && bar`)

**Reply by weirdan** [reply]
`foo && bar` is just `foo; bar` in Nushell, and `foo || bar` is `try { foo } catch { bar }`. 

`foo && bar || baz` would be `try { foo ; bar } catch { baz }` I think. 

Also see https://github.com/nushell/nushell/issues/740#issuecomment-2479809514 which has a bit more in-depth explanation (but does not mention `try/catch`).

**Reply by Jasha10** [reply]
very nice, thanks @weirdan!

**Reply by Jasha10** [reply]
The follow-up question is: what's an ergonomic way to wrap an external command so that we get a nushell error if the exit code is nonzero? (Without having to define a nushell command for each external command).

Here's what I've cooked up:
```nu
# Wrap the output from an external command so that
# an error is raised if the command fails
def wrap [] {
    complete  | do {
        let in_ = $in
        if $in_.exit_code != 0 {
            error make {msg: "External command failed"}
        }
        $in_.stdout
    }
}

```

Here's a demo in action:
```nu
# example with command that succeeds:
$ bash -c 'echo 123' | wrap
123

# example with command that fails:
$ bash -c 'false' | wrap
Error:   × External command failed
     ╭─[/Users/jasha/.config/nushell/config.nu:996:13]
 995 │             }
 996 │             error make {msg: $msg}
     ·             ─────┬────
     ·                  ╰── originates from here
 997 │         }
     ╰────
```

**Reply by Jasha10** [reply]
The above works for my use-case. Strange that I can't seem to capture `stderr`. The record returned by `complete` only contains `exit_code` and `stdout`.

**Reply by fdncred** [reply]
I wonder if you can get `try/catch` to work for you because we recently updated it to be able to pull out different types of error messages.
```nushell
❯ try { bash -c 'echo 123' } catch {|err| $err}
123
❯ try { bash -c 'false' } catch {|err| $err}
╭───────────┬──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│ msg       │ External command had a non-zero exit code                                                                                │
│ debug     │ NonZeroExitCode { exit_code: 1, span: Span { start: 333663, end: 333667 } }                                              │
│ raw       │ NonZeroExitCode { exit_code: 1, span: Span { start: 333663, end: 333667 } }                                              │
│ rendered  │ nu::shell::non_zero_exit_code                                                                                            │
│           │                                                                                                                          │
│           │   × External command had a non-zero exit code                                                                            │
│           │                                                                                                                          │
│ json      │ {"msg":"External command had a non-zero exit code","labels":[{"text":"exited with code                                   │
│           │ 1","span":{"start":333663,"end":333667}}],"code":"nu::shell::non_zero_exit_code","url":null,"help":null,"inner":[]}      │
│ exit_code │ 1                                                                                                                        │
╰───────────┴──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
```
Based off of this PR https://github.com/nushell/nushell/pull/14082

**Reply by Jasha10** [reply]
Very nice -- thank you @fdncred! This clears up some confusion I had.

**Comment by tennox** [comment]
So, the 'best' way to have an error printed when a command fails is this?

```nushell
try {
  external-command
} catch { |err|
  print -e $err.rendered
}
```

**Reply by tennox** [reply]
Moved to https://github.com/nushell/nushell/issues/7153#issuecomment-2678470451

didn't see this thread is about _between_ two commands.
